File: AI/A4.py
# N-Queen Problem in Python

ans = []


def printSolutions():
    print("Total solutions:", len(ans))

    for config in ans:
        for row in config:
            print(row)
        print()


def isSafe(board, row, col, n):
    # Check column
    for i in range(n):
        if board[i][col] == "Q":
            return False

    # Check left diagonal
    i = row
    j = col

    while i >= 0 and j >= 0:
        if board[i][j] == "Q":
            return False
        i -= 1
        j -= 1

    # Check right diagonal
    i = row
    j = col

    while i >= 0 and j < n:
        if board[i][j] == "Q":
            return False
        i -= 1
        j += 1

    return True


def saveBoard(board, n):
    config = []

    for i in range(n):
        row = ""
        for j in range(n):
            row += board[i][j]
        config.append(row)
    ans.append(config)


def nQueens(board, row, n):

    if row == n:
        saveBoard(board, n)
        return

    for j in range(n):
        if isSafe(board, row, j, n):
            board[row][j] = "Q"
            nQueens(board, row + 1, n)
            board[row][j] = "."


def main():
    n = int(input("Enter the size of nxn matrix: "))

    board = []

    for i in range(n):
        temp = []
        for j in range(n):
            temp.append(".")
        board.append(temp)

    nQueens(board, 0, n)
    printSolutions()


main()

File: AI/A4.java

// n-queen
import java.util.*;

public class A4 {
    static ArrayList<ArrayList<String>> ans = new ArrayList<>();

    static void printSolutions() {
        System.out.println("Total solutions: " + ans.size());
        for (ArrayList<String> config : ans) {
            for (String row : config) {
                System.out.println(row);
            }
            System.out.println();
        }
    }

    static boolean isSafe(String board[][], int row, int col, int n) {
        // Check row
        for (int i = 0; i < n; i++) {
            if (board[row][i].equals("Q"))
                return false;
        }
        // Check column
        for (int i = 0; i < n; i++) {
            if (board[i][col].equals("Q"))
                return false;
        }
        // Check left diagonal
        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j].equals("Q"))
                return false;
        }
        // Check right diagonal
        for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
            if (board[i][j].equals("Q"))
                return false;
        }
        return true;
    }

    static void saveBoard(String board[][], int n) {
        ArrayList<String> config = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String row = "";
            for (int j = 0; j < n; j++) {
                row = row + board[i][j];
            }
            config.add(row);
        }
        ans.add(config);
    }

    static void nQueens(String board[][], int row, int n) {
        if (row == n) {
            saveBoard(board, n);
            return;
        }
        for (int j = 0; j < n; j++) {
            if (isSafe(board, row, j, n)) {
                board[row][j] = "Q";
                nQueens(board, row + 1, n);
                board[row][j] = ".";
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter the size of nxn matrix: ");
        int n = sc.nextInt();

        String board[][] = new String[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = ".";
            }
        }

        nQueens(board, 0, n);
        printSolutions();
        sc.close();
    }
}
